refactor(linter/yoda): simplify code

overlookmotel opened the pull request:
Follow-on after #7679.

Simplify `do_diagnostic_with_fix`, in particular the search for the operator. Also reduce `span()` calls, as they have a cost.

overlookmotel commented:
As discussed in https://github.com/oxc-project/oxc/pull/7679#issuecomment-2523025745.

camc314 commented on crates/oxc_linter/src/rules/eslint/yoda.rs:
```diff
@@ -266,61 +266,53 @@ fn is_not_yoda(expr: &BinaryExpression) -> bool {
-
-        let operator_position_start = search_start_position + operator_position_start[0].0 as u32;
+        let (operator_start, operator_end) = str_between_left_and_right
+            .as_bytes()
```
is there any risk of this causing issues with multi byte chars?
multi byte chars are almost always the cause of our panics 🫠

overlookmotel replied:
I don't think so. `operator_str` is always ASCII, so if `chunk == operator_str.as_bytes()` then `pos_start` and `pos_end` must be on UTF8 character boundaries (as they're the start and end of `operator_str`.

Multibyte unicode characters never contain bytes in the ASCII range (0-127). So e.g. `=` can't appear in the middle of a multi-byte sequence.

overlookmotel replied:
Incidentally, same goes for multi-byte characters. No 2 bytes of a 3-byte character can be a valid 2-byte character. So there's never any ambiguity.

camc314 reviewed (approved):
💪

Boshen commented:
### Merge activity

* **Dec 16, 11:40 AM EST**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc) once it meets the requirements.

overlookmotel committed:
refactor(linter/yoda): simplify code (#7941)

Follow-on after #7679.

Simplify `do_diagnostic_with_fix`, in particular the search for the operator. Also reduce `span()` calls, as they have a cost.
